sbi/inference/snpe/snpe_c.py

```diff
@@ -152,13 +180,124 @@ def __call__(
         kwargs = del_entries(locals(), entries=("self", "__class__", "num_atoms"))
 
+        if proposal is not None:
+            self.use_non_atomic_loss = (
+                isinstance(proposal.net._distribution, mdn)
+                and isinstance(self._posterior.net._distribution, mdn)
+                and (
+                    isinstance(self._prior, utils.BoxUniform)
+                    or isinstance(self._prior, mvn.MultivariateNormal)
+                )
+            )
+
+            algorithm = "non-atomic" if self.use_non_atomic_loss else "atomic"
+            print(f"Using SNPE-C with {algorithm} loss")
+
+            if self.use_non_atomic_loss:
+                self._set_state_for_mog_proposal()
+
         return super().__call__(**kwargs)
 
+    def _set_state_for_mog_proposal(self) -> None:
+        """
+        Set state variables that are used at every training step of non-atomic SNPE-C.
+
+        Three things are computed:
+        1) Check if z-scoring was requested. To do so, we check if the `_transform`
+            argument of the net had been a `CompositeTransform`. See pyknos mdn.py.
+        2) Define a (potentially standardized) prior. It's standardized if z-scoring
+            had been requested.
+        3) Compute (Precision * mean) for the prior. This quantity is used at every
+            training step if the prior is Gaussian.
+        """
+
+        self.z_score_theta = isinstance(
+            self._posterior.net._transform, CompositeTransform
+        )
+
+        self._set_maybe_z_scored_prior()
+
+        if isinstance(self._maybe_z_scored_prior, mvn.MultivariateNormal):
+            self.prec_m_prod_prior = torch.mv(
+                self._maybe_z_scored_prior.precision_matrix,
+                self._maybe_z_scored_prior.loc,
+            )
+
+    def _set_maybe_z_scored_prior(self) -> None:
+        r"""
+        Compute and store potentially standardized prior (if z-scoring was requested).
+
+        The proposal posterior is:
+        $pp(\theta|x) = 1/Z * q(\theta|x) * prop(\theta) / p(\theta)$
+
+        Let's denote z-scored theta by `a`: a = (theta - mean) / std
+        Then pp'(a|x) = 1/Z_2 * q'(a|x) * prop'(a) / p'(a)$
+
+        The ' indicates that the evaluation occurs in standardized space. The constant
+        scaling factor has been absorbed into Z_2.
+        From the above equation, we see that we need to evaluate the prior **in
+        standardized space**. We build the standardized prior in this function.
+
+        The standardize transform that is applied to the samples theta does not use
+        the exact prior mean and std (due to implementation issues). Hence, the z-scored
+        prior will not be exactly have mean=0 and std=1.
+        """
+
+        if self.z_score_theta:
+            scale = self._posterior.net._transform._transforms[0]._scale
+            shift = self._posterior.net._transform._transforms[0]._shift
+
+            estim_prior_std = 1 / scale
```

Comment:
I'll make more comments.

But it's actually correct as it is now. `shift=-t_mean / t_std` in `sbiutils.py`. Hence, `t_mean=-shift*t_std=-shift*scale`

Reply:
Ah, I see. what you actually mean is 
"Hence, `t_mean=-shift*t_std=-shift / scale`" 

because `t_std = 1 / scale`. Got it. 

But then, maybe write: 
```
# Following the definition of the scale and shift in `standardizing_transform` in `sbiutils`.
estim_prior_std = 1 / scale
estim_prior_mean = - shift * estim_prior_std
```